### PR TITLE
Allow `rails new . --master` to point to master branch

### PR DESCRIPTION
**Pull request by Schwad**
## Background

Currently the `rails new` generator supports options `--dev` and `--edge`.

`--dev` points to one’s local rails setup and `--edge` points to the latest stable Rails branch.

However, in the Rails community we often think of the ‘edge’ as the latest merged master. We can see as much in Shopify’s fantastic article showing how they point to the latest master version of Rails: https://engineering.shopify.com/blogs/engineering/living-on-the-edge-of-rails

alt

I would originally have recommended changing ‘edge’ to point to master, as it has tripped me up on numerous occasions now.

## Change

I would rather add the desired functionality than change the current edge functionality.

Therefore this PR adds the `--master` flag.

It is intended to do exactly as `edge` does, simply with the change that it points to Rails’ master branch instead of latest stable (i.e. 6_0_stable)

## Okay, but really why?

Whenever developers see a new feature in Rails, such as the recently committed horizontal sharding (which prompted my PR) #38531 , we would love to spin up a new Rails App to try out these new features. ❤️

This command would allow us to jump right in with no configuration and offer our testing of the features, even those of us who aren’t able to spend a lot of time on OSS support. 🙏

### CODE REVIEW

**Comment by Schwad** [comment]
Tested and squashed. 

Output is:

`gem 'rails', github: "rails/rails", branch: "master"`

File: railties/lib/rails/generators/app_base.rb [hunk 1]
```diff
@@ -90,6 +90,8 @@ def self.add_shared_options_for(name)
 
         class_option :edge,                type: :boolean, default: false,
                                            desc: "Set up the #{name} with Gemfile pointing to Rails repository"
+        class_option :master,              type: :boolean, default: false,
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/app_base.rb]:
```suggestion

        class_option :master,              type: :boolean, default: false,
```

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

**Comment by kaspth** [comment]
Love this, great idea!

**Comment by rafaelfranca** [comment]
I backported to 6.0. We usually don't do this, but I think it is worth to allow people to try the awesome things we are building easily.

File: railties/lib/rails/generators/app_base.rb [hunk 2]
```diff
@@ -312,7 +319,7 @@ def assets_gemfile_entry
       def webpacker_gemfile_entry
         return [] if options[:skip_javascript]
 
-        if options.dev? || options.edge?
+        if options.dev? || options.edge? || options.master?
```
Comment (kaspth) [inline railties/lib/rails/generators/app_base.rb]: Think there's a separate PR in extracting this? We probably aren't likely to change this now, but still think it makes sense to make the code more readable. Maybe `pull_from_github?` would do?

**Comment by matthewd** [comment]
This seems moderately inadvisable to me:

We're generating a Gemfile that points to master, but the app is still being generated by whichever stable release you ran the command in.

**Comment by kaspth** [comment]
Hm, right, you'd essentially need to have a clone of Rails somewhere already to generate from ala `~/code/rails/exe/rails new --master`, which is how I usually do it. Not sure if it's within the `new` commands' scope to help with git clones, though it could potentially be neat.

**Comment by matthewd** [comment]
> you'd essentially need to have a clone of Rails somewhere already to generate from ala `~/code/rails/exe/rails new --master`

Right... and if you're running from master, `--edge` already does that. It only uses the corresponding stable branch when run from a release build.

**Comment by Schwad** [comment]
Thank you for the thoughts @matthewd . You raise some great points. 

* Yes, if you ran this command from say Rails 6.0.3, it would generate a 6.0.3 app but pointing to a `master` Gemfile
* At the moment, unless there is a massive gulf in versions (which we'd be 1-2 years away from at least), I'd still say this would work for testing at least 80% of the new features that have been committed ( [like I've done with horizontal sharding](https://weblog.rubyonrails.org/2020/3/1/this-week-in-rails-horizontal-sharding-gzip-schema-cache-database-rake-tasks/) )
* I find it helpful that you noted if I went to the trouble of already bringing in the latest Rails to my machine, that `--edge` would default point to master? (I haven't tested but it sounds right to me)

So my question here is - how do we stomp out the edge cases with this feature?

## The easy way

I could whip up a PR to flag this as 'experimental', or make some other note about the versioning issue. That you *may* get unexpected behavior, etc. 

## The fun way

We could have a look at this codebase and figure out how we could replicate the full master Rails behavior, whilst just using the `--master` flag. Mind you I've only written a handful of custom generators in my life and not architected the guts of the current Rails setup, but you'll get my drift:

1. Template, to force master Rails

```ruby
#=> railties/lib/rails/generators/rails/app/templates/GemfileRailsMaster.tt
source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby <%= "'#{RUBY_VERSION}'" -%>

gem 'rails', github: "rails/rails", branch: "master"
```

2. Install master rails, and generate the brand new app off of master Rails

```ruby
copy_file 'GemfileRailsMaster.tt', 'Gemfile'
puts ('*' * 10) + "\n"
puts 'Pulling down rails:master to generate the new rails application...'
puts "\n" + ('*' * 10)
system 'bundle install'
puts ('*' * 10) + "\n"
puts 'Generating rails application off of latest master commit...'
puts "\n" + ('*' * 10)
system 'bundle exec rails new . --edge'
```

**Comment by rafaelfranca** [comment]
I think if someone wants to use the `--master` options that are ok with the Rails X generated app not working 100% with the Rails X+1 dependency out-of-box. Yeah, some configuration may be different, some repositories in the wrong place. But that is the price of living in the edge.

`rails new --master` doing "install master rails, and generate the brand new app off of master Rails" is an alternative. But I think that would be a totally different command. I just don't have a good name for it.

**Comment by kaspth** [comment]
Hm, my problem with that is that people would open issues or not know how to diss out what’s going on. Could be worth adding a warning that this is a sharp knife etc.

Clearly, that other command would be `rails yolo` 😄

**Comment by matthewd** [comment]
@rafaelfranca  it's not just the risk of _in_-compatibility -- everything that can will be running in _compatibility_ mode, instead of using the new-version defaults. "master as upgraded from the existing stable release" is a valuable and interesting thing to test, but I don't think it's what people would expect to be experiencing here.

@Schwad "it won't be a problem for a year or two" isn't a super-compelling argument for me.. especially because when it does become an issue, we'll have to effective way to fix it retroactively. I do like the sound of the fun way, using a minimal gemfile to bootstrap a master clone, and then handing over to that instance to do the actual work.


@rafaelfranca maybe it's in between: `rails new --master` same command (because the caller doesn't ultimately care about the detail of which copy is doing the generating, they just want a thing that works), but capture the `--master` up in [the command](https://github.com/rails/rails/blob/82d7e3e21708dcb21b47198f996cf3ec9f372612/railties/lib/rails/commands/application/application_command.rb) before it invokes the usual generator?

Obviously anything of this form still has a degree of danger, because we're defining a protocol by which an older Rails version can communicate forward through time to all future Rails versions, so we'll want to think carefully about exactly what command it runs in the new app context -- maybe it's a special undocumented command, rather than a standard `rails new` invocation.


> if I went to the trouble of already bringing in the latest Rails to my machine, that --edge would default point to master?

Yeah... master:

https://github.com/rails/rails/blob/82d7e3e21708dcb21b47198f996cf3ec9f372612/railties/lib/rails/generators/app_base.rb#L287

(For git, no branch == default branch == master)

6-0-stable:

https://github.com/rails/rails/blob/5a0a56d20b0c2c26077fe18c7b5c965107d8be6e/railties/lib/rails/generators/app_base.rb#L287

That incidentally really feels like it should be a constant or something, and coming from somewhere a little more authoritative. 😅🤷🏻‍♂️

**Comment by Schwad** [comment]
Thanks for the feedback everyone, very helpful. :)

 Also, cheers to @roramirez for helping us out with some specs! 🎉

@matthewd : I am attempting to put together an initial pass of the "fun" version [here](https://github.com/Schwad/rails/commits/master).

The way I'm testing it is by locally bundling this gemfile:

```ruby
source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby "2.6.5"

gem 'rails', github: "schwad/rails", branch: "master"
```

Then simply running `bundle exec rails . --master`, and using `puts` debugging where I can for the time being. 

At the moment I'm trying to capture `--master` as early in the chain as possible, use our built-in generators (somehow) to write my template and rebundle and redirect down the `--edge` path. 

Obviously at the moment it's a bit of trial and error, and I've not opened an official PR yet because I am unable to get it working yet. 

I have a freelancer [tuple](https://tuple.app/) license (shoutout to @r00k !) so can pair with anybody on this. If anyone has some time and would be interested to pair with me on this next week let me know. I'm on UK time.

**Comment by Schwad** [comment]
> Obviously anything of this form still has a degree of danger, because we're defining a protocol by which an older Rails version can communicate forward through time to all future Rails versions, so we'll want to think carefully about exactly what command it runs in the new app context -- maybe it's a special undocumented command, rather than a standard rails new invocation.

Alternatively, if my 'fun' pass doesn't make it (though I would still like to give the PR a go because I like the `--master` approach) there could be a straight up `rails yolo` (per @kaspth ) or explicit `rails generate_brand_spanking_new_app_off_master`, which we could 'document' in the community but maybe not explicitly in the first class guides.

**Comment by rafaelfranca** [comment]
I reverted the 6-0-stable backport of this new option while we come up with an implementation that satisfies our requirements. I'm also considering reverting the change on master.  What do you think?

**Comment by Schwad** [comment]
I’ll defer to core on the master branch, but when is 6-1 due, April? 

If so my feeling is that it currently **does** work - albeit as a “sharp tool”. 

That said I’m happy to help make it a more robust tool between now and 6-1 release as well.

File: railties/lib/rails/generators/app_base.rb (hunk 2 above)
Comment (Schwad) [reply railties/lib/rails/generators/app_base.rb]: Oooh good point. I love extracting compound conditionals. After I wrap up #38761 I'll definitely have a dig at that :)

**Pull request merged**